### PR TITLE
Load FOCS scripts from directory

### PR DESCRIPTION
**Pull request by dbenage-cx**
Allows a directory, with multiple files/dirs in place of a regular file for parsing.

This is posted more for design discussion/review before more work.

For now, if the parse file is a directory then all contained files are read for parsing and shared_macros.txt is included.

default/techs.txt was split by category as a simple example of usage

### CODE REVIEW

**Comment by geoffthemedio** [comment]
Why?

**Comment by adrianbroher** [comment]
> Why?

Because.

Really, it is a good idea to allow splitting up the FOCS declarations.  This would enable to change single entities without touching the whole description, reducing editing collisions between a custom variant and the freeorion release variant.

**Comment by adrianbroher** [comment]
Although I think the file naming conventions and the way it is implemented isn't well thought through.

**Comment by dbenage-cx** [comment]
> Why?

I have found it easier to edit in the past and thought it may be a good addition here.

> Although I think the file naming conventions and the way it is implemented isn't well thought through.

Fleshed this out a little more and made a post in 'Other Game Design' forum as that looks to be a favored discussion forum.  I chiefly wanted to provide an example here, potentially gaining some experience in the process even if the small bits of code are scrapped.

**Comment by adrianbroher** [comment]
Don't merge master.

PRs like that will always be rejected, regardless of content.  If you need to reflect master changes rebase the PR.

**Comment by dbenage-cx** [comment]
Will do.

Assuming this would not be resolved by doing a rebase now, and that this becomes acceptable: 
Would it be better to delete this branch and start with a clearer commit history? (if so, I would like to make the actual file splitting into a seperate PR, to make the changes easier to follow)

**Comment by adrianbroher** [comment]
> Assuming this would not be resolved by doing a rebase now, and that this becomes acceptable.

No fear: (interactive) rebasing is the right tool for that.  I do this all the time (rebasing to split, stash, rename and reapply commit history to master) and it works just fine.

> Would it be better to delete this branch and start with a clearer commit history?

There is no need to create a new PR, just force push to your GitHub branch. GitHub will just update the PR accordingly.

> if so, I would like to make the actual file splitting into a seperate PR, to make the changes easier to follow

IMO this would be harder to follow.  Having all the related commits would it make easier to test, because testers are not forced to do the splitting themselves or having to cherry pick commits.  However the user dir related changes should not part of this PR.  I know this is probably your main concern for doing this, but splitting up the files and selecting the proper files from the installation/a mod/user directory are two different topics and can be tackled one after another.

**Comment by adrianbroher** [comment]
Will do an in depth review the next days.

**Comment by dbenage-cx** [comment]
Monster fleet plans have an issue of only keeping the last loaded entry, on my todo list but they may need to stay as a single file.

I still need to look into keymaps before bringing in the file splits and updating doc.

Keeping removal of tests and changes affecting it as separate commits in case a revert is needed later.

**Comment by geoffthemedio** [comment]
When I pull and build this pull request, upon startup, freeorion.exe hangs, not responding. In the log file, I see the last line:

```
2016-03-04 18:50:10.074636 [error] Client : Parse.cpp:539 : Unable to open data file scripting/techs/Categories.inf
```

Looking into the code, you're not handling paths portably. In particular, passing in something like

```
ListScripts("scripting/ship_hulls");
```

to a function that takes a boost::path won't work. You'll possibly need to concatenate directories and file names using the / operator as in

```
parse::ship_parts(GetResourceDir() / "ship_parts.txt", m_parts);
```

**Comment by dbenage-cx** [comment]
Rebased

**Comment by Vezzra** [comment]
@dbenage-cx, in case you missed @geoffthemedio's [comment above](https://github.com/freeorion/freeorion/pull/481#issuecomment-192379267), you need to fix the way you handle paths.

**Comment by dbenage-cx** [comment]
Sorry, he had made the pull just before I corrected that and rebased it.  [Categories.inf changed from .focs.txt](https://github.com/dbenage-cx/freeorion/commit/d7d33e273371f34b1006fd58443c2daf4a365bc0#diff-5f69595021dfb8b31275cf73e9ec661aR260), just to make sure there is no issue loading a tech before.

**Comment by geoffthemedio** [comment]
Doesn't look fixed to me... still has calls like

```
ListScripts("scripting/ship_parts")
```

**Comment by dbenage-cx** [comment]
[ListScripts](https://github.com/dbenage-cx/freeorion/commit/d7d33e273371f34b1006fd58443c2daf4a365bc0#diff-6ce52f5b90a662818a62fbd02ba6de8fR431) takes a relative or absolute path, calling [ListDir](https://github.com/dbenage-cx/freeorion/commit/d7d33e273371f34b1006fd58443c2daf4a365bc0#diff-79cd8680ec17b985fc82fde5e9b54368R370).

I can put the calls to GetResourceDir() back in there if it will help clear things up, just saw an opportunity to consolidate.

**Comment by adrianbroher** [comment]
> Doesn't look fixed to me... still has calls like

This is fine as is.  ListScripts should be in charge of resolving resources from multiple possible source directories (think install dir, mod dir, user dir).

**Comment by geoffthemedio** [comment]
My concern was writing "directory/otherdirectory" or similar, using the / in a string, as boost path seems to have issues unless you write it like "directory" / "otherdirectory".

But it seems to work now, so the problem was apparently not what I thought.

**Comment by adrianbroher** [comment]
> My concern was writing "directory/otherdirectory" or similar, using the / in a string, as boost path seems to have issues unless you write it like "directory" / "otherdirectory".

This adheres to the generic pathname format used by boost filesystem.

http://www.boost.org/doc/libs/1_60_0/libs/filesystem/doc/reference.html#generic-pathname-format
http://www.boost.org/doc/libs/1_53_0/libs/filesystem/doc/reference.html#path-Conversions

**Comment by geoffthemedio** [comment]
my concern was that it wouldn't interpret it as a generic path format correctly on Windows

**Comment by dbenage-cx** [comment]
Will put in path.make_preferred() in ListDir

**Comment by geoffthemedio** [comment]
As I wrote above, it already appears to work fine.

**Comment by dbenage-cx** [comment]
Ok, leaving it alone :) rebased to keep pace

File: default/scripting/buildings/colonies/col_bld_gen.py [hunk 1]
```diff
@@ -203,12 +205,13 @@
...
-        sp_id = species[0]
-        sp_name = sp_id.split("_", 1)[1]
-        sp_desc_name = species[1]
-        sp_graphic = species[2]
+for species in species_list:
+    sp_id = species[0]
+    sp_name = sp_id.split("_", 1)[1]
+    sp_desc_name = species[1]
+    sp_graphic = species[2]
+    sp_filename = sp_id + ".fos.txt"
```
Comment (Morlic-fo) [inline default/scripting/buildings/colonies/col_bld_gen.py]: Shouldn't this be `.focs.txt`?

**Comment by dbenage-cx** [comment]
will setup a rough guide of 120 columns.
want these squashed in? (standalone script and formatting)

**Comment by geoffthemedio** [comment]
squashing is unnecessary

**Comment by geoffthemedio** [comment]
Do you have a "real" name for the credits?

**Comment by dbenage-cx** [comment]
D. Benage

**Comment by Vezzra** [comment]
Thanks, added you to the credits: 21632708635510f4cf303718a1c02f1f3395456f

**Pull request merged**